apollo-router/src/plugins/demand_control/cost_calculator/schema_aware_response.rs

```diff
@@ -49,38 +50,83 @@ impl<'a> SchemaAwareResponse<'a> {
     }
 }
 
+pub(crate) trait Visitor {
+    fn visit(&self, value: &TypedValue) {
+        match value {
+            TypedValue::Null => self.visit_null(),
+            TypedValue::Bool(ty, f, b) => self.visit_bool(ty, f, b),
+            TypedValue::Number(ty, f, n) => self.visit_number(ty, f, n),
+            TypedValue::String(ty, f, s) => self.visit_string(ty, f, s),
+            TypedValue::List(ty, f, items) => self.visit_array(ty, f, items),
+            TypedValue::Object(ty, f, children) => self.visit_object(ty, f, children),
+            TypedValue::Root(children) => self.visit_root(children),
+        }
+    }
+    fn visit_field(&self, _value: &TypedValue) {}
+    fn visit_null(&self) {}
+    fn visit_bool(&self, _ty: &NamedType, _field: &Field, _value: &bool) {}
+    fn visit_number(&self, _ty: &NamedType, _field: &Field, _value: &serde_json::Number) {}
+    fn visit_string(&self, _ty: &NamedType, _field: &Field, _value: &str) {}
+
+    fn visit_array(&self, _ty: &NamedType, _field: &Field, items: &[TypedValue]) {
+        for value in items.iter() {
+            self.visit_array_element(value);
+            self.visit(value);
+        }
+    }
+    fn visit_array_element(&self, _value: &TypedValue) {}
+    fn visit_object(
+        &self,
+        _ty: &NamedType,
+        _field: &Field,
+        children: &HashMap<String, TypedValue>,
+    ) {
+        for value in children.values() {
+            self.visit_field(value);
+            self.visit(value);
+        }
+    }
+    fn visit_root(&self, children: &HashMap<String, TypedValue>) {
+        for value in children.values() {
+            self.visit_field(value);
+            self.visit(value);
+        }
+    }
+}
+
 #[derive(Debug)]
 pub(crate) enum TypedValue<'a> {
     Null,
-    Bool(&'a Field, &'a bool),
-    Number(&'a Field, &'a serde_json::Number),
-    String(&'a Field, &'a str),
-    Array(&'a Field, Vec<TypedValue<'a>>),
-    Object(&'a Field, HashMap<String, TypedValue<'a>>),
+    Bool(&'a NamedType, &'a Field, &'a bool),
```

Comment:
Need to add test for NamedType

Reply:
[60fa19a](https://github.com/apollographql/router/pull/5215/commits/60fa19aae60c26bd382e6ed01d7d3df72e6efd39)